### PR TITLE
Train architecture from Ounkomol et. al. to predict cropped nuclei from dapi to gold images

### PR DESCRIPTION
**Pull request by MattsonCam**
This will be a two-step pr. The first step is adding the model code, so any adjustments can be made prior to training. After reviewing any pr comments I will then push the trained model along with model artifacts, images, and a readme. This pr trains the model architecture created by Ounkomol et. al. Some noteworthy comments:

- Not all imports are at the beginning of the script, as it is cleaner, in my opinion, to defined the root git directory before some of the import statements. This eliminates the need to define relative paths in some of these import statements.
- I defined the random seeds before the inputs and outputs, and would be interested to hear your thoughts on this decision.
- No comments were created inside of the model architecture code, instead, I left this code intact from the authors (https://doi.org/10.1038/s41592-018-0111-2)
-  I decided to not create copies of transformed images to reduce dataset complexity, as I do not anticipate using these cropped images for training other models. Instead, I transform previously cropped images at runtime.
- I save the images as .png files. I would be interested to hear thoughts on data formats.

Feel free to add any constructive comments when reviewing.

### CODE REVIEW

File: 1.develop_vision_models/fnet_model/nbconverted/fnet_predict_nuclear_speckle_stain.py
Comment (MikeLippincott) [inline 1.develop_vision_models/fnet_model/nbconverted/fnet_predict_nuclear_speckle_stain.py]: Consider resolving these paths using pathlib

File: 1.develop_vision_models/fnet_model/nbconverted/fnet_predict_nuclear_speckle_stain.py
Comment (MikeLippincott) [inline 1.develop_vision_models/fnet_model/nbconverted/fnet_predict_nuclear_speckle_stain.py]: make sure these are run in part 2, but looking good so far!

File: 1.develop_vision_models/fnet_model/nbconverted/fnet_predict_nuclear_speckle_stain.py
Comment (MikeLippincott) [inline 1.develop_vision_models/fnet_model/nbconverted/fnet_predict_nuclear_speckle_stain.py]: Consider adding docstrings and type hints here

File: 1.develop_vision_models/fnet_model/nbconverted/fnet_predict_nuclear_speckle_stain.py
Comment (MikeLippincott) [inline 1.develop_vision_models/fnet_model/nbconverted/fnet_predict_nuclear_speckle_stain.py:318]: consider writing to parquet or database?

File: 1.develop_vision_models/models/fnet_nn_2d.py
Comment (MikeLippincott) [inline 1.develop_vision_models/models/fnet_nn_2d.py:22]:
```suggestion
        n_in_channels - (int) number of channels for input.
```

File: 1.develop_vision_models/models/fnet_nn_2d.py
Comment (MikeLippincott) [inline 1.develop_vision_models/models/fnet_nn_2d.py:24]: I am not sure what this means

File: 1.develop_vision_models/models/fnet_nn_2d.py
Comment (MikeLippincott) [inline 1.develop_vision_models/models/fnet_nn_2d.py:9]: Not sure what `mult_chan` is? same with `depth`. Depth of what? I think some docstring documentation would not hurt here.

File: 1.develop_vision_models/fnet_model/nbconverted/fnet_predict_nuclear_speckle_stain.py
Comment (MikeLippincott) [inline 1.develop_vision_models/fnet_model/nbconverted/fnet_predict_nuclear_speckle_stain.py]: Could these params be defined in a single nested dict or in say a `toml` file?

File: 1.develop_vision_models/ModelTrainer.py
Comment (MikeLippincott) [inline 1.develop_vision_models/ModelTrainer.py:23]: A docstring would be nice here. I have no clue what patience is?

File: 1.develop_vision_models/ModelTrainer.py
Comment (MikeLippincott) [inline 1.develop_vision_models/ModelTrainer.py:6]: If this is from the original fnet paper consider citing

**Review by MikeLippincott — APPROVED** [review]
LGTM. I left specific code comments that should be addressed. I am excited and looking forward to seeing part two of this! Good work!

File: 1.develop_vision_models/fnet_model/nbconverted/fnet_predict_nuclear_speckle_stain.py
Comment (MattsonCam) [reply 1.develop_vision_models/fnet_model/nbconverted/fnet_predict_nuclear_speckle_stain.py]: Resolved!

File: 1.develop_vision_models/fnet_model/nbconverted/fnet_predict_nuclear_speckle_stain.py
Comment (MattsonCam) [reply 1.develop_vision_models/fnet_model/nbconverted/fnet_predict_nuclear_speckle_stain.py:318]: Could be useful, Greg wanted me to store it in a tidy csv format for now

File: 1.develop_vision_models/fnet_model/nbconverted/fnet_predict_nuclear_speckle_stain.py
Comment (MattsonCam) [reply 1.develop_vision_models/fnet_model/nbconverted/fnet_predict_nuclear_speckle_stain.py]: Good idea, I added type hints here. I think together the code and the one function description summarize what is happening here. e.g. the model computes the output given the input `output = model(single_input)`

File: 1.develop_vision_models/models/fnet_nn_2d.py
Comment (MattsonCam) [reply 1.develop_vision_models/models/fnet_nn_2d.py:24]: Those are details of the author's unet-inspired model. I left these comments and the remainder of the file intact

File: 1.develop_vision_models/models/fnet_nn_2d.py
Comment (MattsonCam) [reply 1.develop_vision_models/models/fnet_nn_2d.py:22]: Right now I'm leaving this file from the original author intact

File: 1.develop_vision_models/models/fnet_nn_2d.py
Comment (MattsonCam) [reply 1.develop_vision_models/models/fnet_nn_2d.py:9]: It defines the number of recursive calls. See above comments

File: 1.develop_vision_models/fnet_model/nbconverted/fnet_predict_nuclear_speckle_stain.py
Comment (MattsonCam) [reply 1.develop_vision_models/fnet_model/nbconverted/fnet_predict_nuclear_speckle_stain.py]: Probably, I was passing parameters using argparse, but this didn't work when running a jupyter notebook. I could see this being faster for training. On gh though, I think this is more of a preference, but maybe this will change later

File: 1.develop_vision_models/ModelTrainer.py
Comment (MattsonCam) [reply 1.develop_vision_models/ModelTrainer.py:23]: Noted, we also talked about this in-person afterwards. Just to summarize: I think patience is the least familiar term and I think the code would be able to describe what it is. The other terms I think are self-explanatory if you are familiar with deep learning models. I also added a comment describing patience, where the patience instance variable is defined, which hopefully adds some clarity. Just trying not to over-describe here in the spirit of (DRY)!

File: 1.develop_vision_models/ModelTrainer.py
Comment (MattsonCam) [reply 1.develop_vision_models/ModelTrainer.py:6]: These imports aren't from the original fnet paper implementation, nor is `ModelTrainer.py`. Similarly, it might be worth adding a comment where I import the fnet to credit the authors

**Comment by MattsonCam** [comment]
Thanks for the first review @MikeLippincott, requesting the second review

File: 1.develop_vision_models/fnet_model/fnet_validation_images/10046_slide3_A2_M45_CH0_Z09.png
Comment (MikeLippincott) [inline 1.develop_vision_models/fnet_model/fnet_validation_images/10046_slide3_A2_M45_CH0_Z09.png:1]: Consider only saving a few examples of these to github and changing contrast. These are showing up as black boxes currently.

Comment (MikeLippincott) [reply 1.develop_vision_models/fnet_model/fnet_validation_images/10046_slide3_A2_M45_CH0_Z09.png:1]: Consider adding a training PR curve or loss curve, this might be more informative than saving every image prediction.

Comment (MikeLippincott) [reply 1.develop_vision_models/fnet_model/fnet_validation_images/10046_slide3_A2_M45_CH0_Z09.png:1]: Alternativly, you could plot the distribution of your chosen metric and view times from the low, middle and high ends of the distribution to asses how this model is performing.

**Review by MikeLippincott — APPROVED** [review]
LGTM, I would reccommend not adding every image to github, but only a few as examples. Ready to merge after addressing comments.

File: 1.develop_vision_models/fnet_model/fnet_validation_images/10046_slide3_A2_M45_CH0_Z09.png
Comment (MattsonCam) [reply 1.develop_vision_models/fnet_model/fnet_validation_images/10046_slide3_A2_M45_CH0_Z09.png:1]: > Consider only saving a few examples of these to github and changing contrast. These are showing up as black boxes currently.

I think this is because the range is so large


> Consider adding a training PR curve or loss curve, this might be more informative than saving every image prediction.

It could be a nice visual. I might do that with the final nuclear speckle model, or maybe I'll add that later with this model. Right now, I just need a way to compare approaches. I'm also only saving 10 random images of each siRNA "category". There are over 200k images. I think I pushed too many images by re-running the model. I will fix that


> Alternativly, you could plot the distribution of your chosen metric and view times from the low, middle and high ends of the distribution to asses how this model is performing.

That's a possibility, I think randomly sampling images here give a good enough representation of how the tends to perform.

**Comment by MattsonCam** [comment]
Thanks @MikeLippincott, merging now!

**Pull request merged**
